pkg/api/types.go

```diff
@@ -688,36 +691,66 @@ type ServiceStatus struct{}
 
 // ServiceSpec describes the attributes that a user creates on a service
 type ServiceSpec struct {
-	// Port is the TCP or UDP port that will be made available to each pod for connecting to the pods
-	// proxied by this service.
-	Port int `json:"port"`
+	// Required: The list of ports that are exposed.
+	// FIXME: this would be nicer as a map of name -> struct
+	Ports []ServicePort
 
-	// Optional: Supports "TCP" and "UDP".  Defaults to "TCP".
-	Protocol Protocol `json:"protocol,omitempty"`
+	// Either 'Selector' or 'Endpoints' must be set:
 
-	// This service will route traffic to pods having labels matching this selector. If empty or not present,
-	// the service is assumed to have endpoints set by an external process and Kubernetes will not modify
-	// those endpoints.
+	// This service will route traffic to pods having labels matching this
+	// selector.
 	Selector map[string]string `json:"selector"`
 
+	// Use the specified endpoints.
```

Comment:
There is a factoring that we discussed, which I looked at and didn't like (too little ROI):

Make a new TargetSet struct something like:

```
type TargetSetPort struct {
    Role string
    Protocol Protocol
    Port IntOrString
    Count int
}
 
type TargetSet {
    ObjectMeta
    // Choose one of FromSelector or FromEndpoints
    FromSelector map[string]string
    FromEndpoints []net.IP
    Ports []TargetSetPort
}
```

Then reduce ServiceSpec to something like

```
type ServicePort {
    Role string
    Port int
    Count int
}

type ServiceSpec struct {
    Ports []ServicePort
    TargetSet ObjectReference
}
```

I played with it buut I didn't think it added much real value but made the API significantly harder to comprehend.  I'm willing to revisit...